### PR TITLE
Rework command line argument stuff

### PR DESCRIPTION
**Pull request by derrod**
### Description

- Adds new `os_process_args_t` and associated API to construct command line arguments
- Adds `os_process_pipe_create2()` to take advantage of new API
- Migrates `obs-ffmpeg-mux` to the new API

### Motivation and Context

Want a saner API than passing stuff to `/bin/sh` and dying inside while concatenating strings.

### How Has This Been Tested?

Recorded with output that requires ffmpeg muxer.

### Types of changes

- Tweak (non-breaking change to improve existing functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by tt2468** [comment]
Considering the relative generic-ness of the `os_process_args_t` functionality, could this also be used by the frontend to process its own arguments? Seems like it would allow for some code deduplication both in obs-app and obs-websocket.

**Comment by derrod** [comment]
> Considering the relative generic-ness of the `os_process_args_t` functionality, could this also be used by the frontend to process its own arguments? Seems like it would allow for some code deduplication both in obs-app and obs-websocket.

There's no parsing happening here, this is purely about assembling arguments for calling subprocesses. Could perhaps be used for launching the updater[^1] I guess.

[^1]: https://github.com/obsproject/obs-studio/blob/master/UI/update/win-update.cpp#L330

File: libobs/util/pipe-windows.c [hunk 1]
```diff
@@ -145,6 +146,69 @@ os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
...
+
+		if (cmd_line.len)
+			dstr_cat_ch(&cmd_line, ' ');
+		if (needs_quotes)
+			dstr_cat_ch(&cmd_line, '"');
+
+		while (*arg) {
+			if (*arg == '\\')
+				bs_count++;
+			else if (*arg == '"') {
```
Comment (RytoEX) [inline libobs/util/pipe-windows.c:180]: If one `if-else if-else` branch uses curly braces, they must all use curly braces.

**Review by RytoEX — CHANGES_REQUESTED** [review]

File: libobs/util/pipe-posix.c [hunk 2]
```diff
@@ -121,6 +120,23 @@ os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
 	return out;
 }
 
+os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
+					  const char *type)
+{
+	if (!cmd_line)
+		return NULL;
+
+	char *argv[4] = {"sh", "-c", (char *)cmd_line, NULL};
```
[the remark this answers is not in the document]
Comment (derrod) [reply libobs/util/pipe-posix.c:129]: Oh yeah that could probably be removed.

File: libobs/util/pipe.c [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+void os_process_args_destroy(struct os_process_args *args)
+{
+	if (!args)
+		return;
+
+	for (size_t idx = 0; idx < args->arguments.num; idx++)
+		bfree(args->arguments.array[idx]);
+	da_free(args->arguments);
```
Comment (RytoEX) [inline libobs/util/pipe.c:73]: Nit: line break before to clearly separate from the `for` body.

File: libobs/util/pipe.c [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+	va_start(va_args, format);
+	dstr_vprintf(&tmp, format, va_args);
+	da_insert(args->arguments, args->arguments.num - 1, &tmp.array);
+	va_end(va_args);
+}
+
+size_t os_process_args_get_argc(struct os_process_args *args)
+{
+	/* Do not count terminating NULL */
```
Comment (RytoEX) [inline libobs/util/pipe.c:57]:
```suggestion
	/* Do not count terminating NULL. */
```

File: libobs/util/pipe.c [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+	/* Last item in argv must be NULL */
+	char *terminator = NULL;
+	da_push_back(args->arguments, &terminator);
+	return args;
+}
+
+void os_process_args_add_arg(struct os_process_args *args, const char *arg)
+{
+	char *str = bstrdup(arg);
+	/* Insert before NULL list terminator */
```
Comment (RytoEX) [inline libobs/util/pipe.c:39]:
```suggestion
	/* Insert before NULL list terminator. */
```

File: libobs/util/pipe.c [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+struct os_process_args {
+	DARRAY(char *) arguments;
+};
+
+struct os_process_args *os_process_args_create(const char *executable)
+{
+	struct os_process_args *args = bzalloc(sizeof(struct os_process_args));
+	char *str = bstrdup(executable);
+	da_push_back(args->arguments, &str);
+	/* Last item in argv must be NULL */
```
Comment (RytoEX) [inline libobs/util/pipe.c:30]:
```suggestion
	/* Last item in argv must be NULL. */
```

File: libobs/util/pipe-windows.c [hunk 7]
```diff
@@ -145,6 +146,69 @@ os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
...
+os_process_pipe_t *os_process_pipe_create2(const os_process_args_t *args,
+					   const char *type)
+{
+	struct dstr cmd_line = {0};
+
+	/* Convert list to command line as Windows does not have any API that
+	 * allows us to just pass argc/argv */
+	char **argv = os_process_args_get_argv(args);
+
+	/* Based on python subprocess module implementation */
```
Comment (RytoEX) [inline libobs/util/pipe-windows.c:164]:
```suggestion
	/* Based on Python subprocess module implementation. */
```

File: libobs/util/pipe-windows.c [hunk 8]
```diff
@@ -145,6 +146,69 @@ os_process_pipe_t *os_process_pipe_create(const char *cmd_line,
...
+		dstr_cat_ch(str, '\\');
+}
+
+os_process_pipe_t *os_process_pipe_create2(const os_process_args_t *args,
+					   const char *type)
+{
+	struct dstr cmd_line = {0};
+
+	/* Convert list to command line as Windows does not have any API that
+	 * allows us to just pass argc/argv */
```
Comment (RytoEX) [inline libobs/util/pipe-windows.c:161]:
```suggestion
	 * allows us to just pass argc/argv. */
```

**Review by RytoEX — CHANGES_REQUESTED** [review]
Nits. Otherwise, this _seems_ okay at a glance.

**Review by Lain-B — APPROVED** [review]

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 9]
```diff
@@ -297,39 +302,36 @@ static void build_command_line(struct ffmpeg_muxer *stream, struct dstr *cmd,
 		num_tracks++;
 	}
 
-	dstr_init_move_array(cmd, os_get_executable_path_ptr(FFMPEG_MUX));
-	dstr_insert_ch(cmd, 0, '\"');
-	dstr_cat(cmd, "\" \"");
-
-	dstr_copy(&stream->path, path);
```
Comment (kkartaltepe) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c:304]: This was actually used in a bunch of places outside build_command_line for printing the destination path, now starting ffmpeg_muxer outputs just says its recording to null.

Comment (derrod) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c:304]: Ah that's an oversight, I guess we still need to copy to `&stream->path`, should be an easy fix at least.

**Pull request merged**